core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -382,10 +373,9 @@ private boolean tryRehash()
         int newCapacity = toIntExact(newCapacityLong);
 
         // An estimate of how much extra memory is needed before we can go ahead and expand the hash table.
-        // This includes the new capacity for groupAddressByHash, rawHashByHashPosition, groupIdsByHash, and groupAddressByGroupId as well as the size of the current page
-        preallocatedMemoryInBytes = (newCapacity - hashCapacity) * (long) (Long.BYTES + Integer.BYTES + Byte.BYTES) +
-                (long) (calculateMaxFill(newCapacity) - maxFill) * Long.BYTES +
-                currentPageSizeInBytes;
+        // This includes the new capacity for rawHashByHashPosition, groupIdsByHash as well as the size of the current page
+        preallocatedMemoryInBytes = (newCapacity - hashCapacity) * (long) (Integer.BYTES + Byte.BYTES)
+                + Math.max(lastFullPageRetainedSize, currentPageSizeInBytes);
```

Comment:
undo. `currentPageSizeInBytes` represents in-flight input page and not page from `channelBuilders`. Currently built page is already accounted for in `getEstimatedSize`: `currentPageBuilder.getRetainedSizeInBytes()`.

Reply:
doesn't `currentPageSizeInBytes` here respresents that possible extension of `channelBuilders` by a new page?
If not, why do we only account for the in-flight page during rehash?

Reply:
> doesn't currentPageSizeInBytes here respresents that possible extension of channelBuilders by a new page?

no.

> If not, why do we only account for the in-flight page during rehash?

Look at `getEstimatedSize`. We account for preallocated memory (including in-flight page), currently built page and accumulated pages

Reply:
`preallocatedMemoryInBytes` is only set during rehash, in normal processing its 0 so does not include `currentPageSizeInBytes`

Reply:
Ok, but for normal processing input page is fully consumed, no need to account it's memory